### PR TITLE
feat(cli): intro warning (35) support for deferred constraints

### PR DESCRIPTION
**Pull request by Druue**
closes https://github.com/prisma/schema-team/issues/386

### CODE REVIEW

File: packages/migrate/src/__tests__/DbPull.test.ts [hunk 1]
```diff
@@ -1779,66 +1739,68 @@ describe('postgresql null-sorted indices introspection', () => {
...
+      expect(ctx.mocked['process.stdout.write'].mock.calls.join('\n')).toMatchInlineSnapshot(``)
+      expect(ctx.mocked['process.stderr.write'].mock.calls.join('\n')).toMatchInlineSnapshot(``)
     })
-  }
+  })
 
-  describe('postgres partioned 30 - row level security found', () => {
-    const warningCode = 30
-    setupPostgresForWarning(warningCode)
+  describe('postgres partioned 35 - deferred constraint found', () => {
```
Comment (Druue) [inline packages/migrate/src/__tests__/DbPull.test.ts:1797]: This is the test for code 35

File: packages/migrate/src/__tests__/DbPull.test.ts [hunk 2]
```diff
@@ -1621,27 +1622,28 @@ describe('postgresql partitioned tables introspection', () => {
...
+                                                                                                                                                                                                                                // - Model "measurement"
+                                                                                                                                                                                                                                // 
+                                                                                                                                                                        `)
       expect(ctx.mocked['process.stdout.write'].mock.calls.join('\n')).toMatchInlineSnapshot(``)
       expect(ctx.mocked['process.stderr.write'].mock.calls.join('\n')).toMatchInlineSnapshot(``)
     })
   })
 
   describe('postgres partioned 27/2 - multiple partitioned tables found', () => {
+    const stopGap = 'partitioned'
```
Comment (janpio) [inline packages/migrate/src/__tests__/DbPull.test.ts:1640]: This `describe` and the previous one could be wrapped in another `describe` that share a `const StopGap` and `cost warningCode` maybe?

File: packages/migrate/src/__tests__/DbPull.test.ts [hunk 3]
```diff
@@ -1525,25 +1525,25 @@ describe('postgresql views introspection warnings', () => {
...
+                                                                                                                                                                                                            `)
       expect(ctx.mocked['process.stdout.write'].mock.calls.join('\n')).toMatchInlineSnapshot(``)
       expect(ctx.mocked['process.stderr.write'].mock.calls.join('\n')).toMatchInlineSnapshot(``)
     })
   })
 })
 
-describe('postgresql partitioned tables introspection', () => {
+describe('postgresql introspection stopgaps', () => {
   const connectionString = process.env.TEST_POSTGRES_URI_MIGRATE!.replace('tests-migrate', 'tests-migrate-db-pull')
```
Comment (janpio) [inline packages/migrate/src/__tests__/DbPull.test.ts:1544]: Sure that it is ok to use the same connectionstring for all these tests? Not the danger that they cross over each other maybe?

File: packages/migrate/src/__tests__/DbPull.test.ts [hunk 4]
```diff
@@ -1867,8 +1836,10 @@ describe('postgresql row level security introspection', () => {
 
         // *** WARNING ***
         // 
-        // These tables contain row level security, which is not yet fully supported.
-        // - Model "foo"
+        // These primary key, foreign key or unique constraints are using non-default deferring in the database, which is not yet fully supported. Read more: https://pris.ly/d/constraint-deferring
+        // -Model "a", Constraint "foo_key"
+        // -Model "a", Constraint "foo_pkey"
+        // -Model "a", Constraint "a_b_fk"
```
Comment (janpio) [inline packages/migrate/src/__tests__/DbPull.test.ts:1842]: This is missing a space between `-` and `Model`

File: packages/migrate/src/commands/DbPull.ts [hunk 5]
```diff
@@ -468,29 +486,59 @@ ${`Run ${chalk.green(getCommandWithExecutor('prisma generate'))} to generate Pri
           warning.code === 31
         ) {
           message += warning.affected.map((it) => `- Model "${it.model}"`).join('\n')
-        } else if (warning.code === 20) {
+        } else if (
+          // * AffectedModelAndConstraint
+          warning.code === 35
+        ) {
+          message += warning.affected.map((it) => `-Model "${it.model}", Constraint "${it.constraint}"`).join('\n')
```
Comment (janpio) [inline packages/migrate/src/commands/DbPull.ts:493]:
```suggestion
          message += warning.affected.map((it) => `- Model "${it.model}", Constraint "${it.constraint}"`).join('\n')
```

**Review by janpio — APPROVED** [review]
Looks great, but some ideas in comments.

File: packages/migrate/src/commands/DbPull.ts [hunk 6]
```diff
@@ -468,29 +486,59 @@ ${`Run ${chalk.green(getCommandWithExecutor('prisma generate'))} to generate Pri
...
+        } else if (
+          // * AffectedTopLevel
+          warning.code === 20
```
Comment (millsp) [inline packages/migrate/src/commands/DbPull.ts:496]: Could we use a numeric enum for these cases?

```ts
enum Response {
  No = 0,
  Yes = 1,
}
```

**Review by millsp — APPROVED** [review]

File: packages/migrate/src/__tests__/DbPull.test.ts (hunk 2 above)
Comment (jkomyno) [reply packages/migrate/src/__tests__/DbPull.test.ts:1640]: I agree, let's leverage hierarchical `describe` blocks. Done

File: packages/migrate/src/__tests__/fixtures/introspection/postgresql/deferred-constraint-warning-35/setup.sql [hunk 7]
```diff
@@ -0,0 +1,24 @@
...
+
+ALTER TABLE a
+    ADD CONSTRAINT foo_key
+    UNIQUE(foo)
+    DEFERRABLE INITIALLY IMMEDIATE;
+
+ALTER TABLE a
+    ADD CONSTRAINT foo_pkey
+    PRIMARY KEY (id)
+    DEFERRABLE INITIALLY DEFERRED;
```
[the remark this answers is not in the document]
Comment (Druue) [reply packages/migrate/src/__tests__/fixtures/introspection/postgresql/deferred-constraint-warning-35/setup.sql:24]: These should all generate warnings, the default constraint case that we currently support per the notion page:

```sql
ALTER TABLE foo ADD CONSTRAINT foo_exclude
      EXCLUDE (bar WITH =)
      NOT DEFERRABLE INITIALLY IMMEDIATE;
```

[the remark this answers is not in the document]
Comment (pimeys) [reply packages/migrate/src/__tests__/fixtures/introspection/postgresql/deferred-constraint-warning-35/setup.sql:24]: The thing that _will not_ generate warnings is the one @Druue described here. All other combinations will warn.

[the remark this answers is not in the document]
Comment (pimeys) [reply packages/migrate/src/__tests__/fixtures/introspection/postgresql/deferred-constraint-warning-35/setup.sql:24]: Exclude will not (yet) generate warnings due to us not even having a stopgap for them.

File: packages/migrate/src/commands/DbPull.ts (hunk 6 above)
Comment (janpio) [reply packages/migrate/src/commands/DbPull.ts:496]: What do you mean?

File: packages/migrate/src/__tests__/fixtures/introspection/postgresql/deferred-constraint-warning-35/setup.sql (hunk 7 above)
[the remark this answers is not in the document]
Comment (Druue) [reply packages/migrate/src/__tests__/fixtures/introspection/postgresql/deferred-constraint-warning-35/setup.sql:24]: Yep @janpio

File: packages/migrate/src/commands/DbPull.ts (hunk 6 above)
Comment (millsp) [reply packages/migrate/src/commands/DbPull.ts:496]: Explanation over here https://prisma-company.slack.com/archives/C02FNFLDUS3/p1681740382535349?thread_ts=1681401438.067269&cid=C02FNFLDUS3

Comment (Druue) [reply packages/migrate/src/commands/DbPull.ts:496]: Resolved per [slack thread](https://prisma-company.slack.com/archives/C02FNFLDUS3/p1681740524693499?thread_ts=1681401438.067269&cid=C02FNFLDUS3)

**Pull request merged**